### PR TITLE
Cosmos improve request diagnostics to capture ClientStatistics counts / config

### PR DESCRIPTION
**Pull request by moderakh**
This PR adds the following to the cosmos request diagnostics:
```
"clientCfgs": {
	"id": 0,
	"numberOfClients": 1,
	"connCfg": {
		"rntbd": "(cto:PT5S, rto:PT5S, icto:PT0S, ieto:PT1H, mcpe:130, mrpc:30)",
		"gw": "(cps:1000, rto:PT5S, icto:null, p:false)",
		"other": "(ed: true, cs: false)"
	},
	"consistencyCfg": "(consistency: Eventual, mm: true, prgns: [westus1,eastus1])"
}
```

reviewers please note that although many files changed but the change is minimal (passing the ClientContext into DSR and to the constructor of the component working with DSR).

TODO: more cleanup to only include important config in the above

```json
{
	"userAgent": "azsdk-java-cosmos/4.6.0-beta.1 MacOSX/10.15.6 JRE/11.0.6",
	"requestLatencyInMs": 1,
	"requestStartTimeUTC": "2020-09-30T16:06:03.703178Z",
	"requestEndTimeUTC": "2020-09-30T16:06:03.704396Z",
	"connectionMode": "DIRECT",
	"responseStatisticsList": [{
		"storeResult": {
			"storePhysicalAddress": "rntbd://10.37.129.3:10253/apps/DocDbApp/services/DocDbServer2/partitions/a4cb494e-38c8-11e6-8106-8cdcd42c33be/replicas/1p/",
			"lsn": 1731,
			"globalCommittedLsn": -1,
			"partitionKeyRangeId": "1",
			"isValid": true,
			"statusCode": 200,
			"subStatusCode": 0,
			"isGone": false,
			"isNotFound": false,
			"isInvalidPartition": false,
			"requestCharge": 1.0,
			"itemLSN": 1731,
			"sessionToken": "-1#1731",
			"exception": null,
			"transportRequestTimeline": [{
				"eventName": "created",
				"durationInMicroSec": "8",
				"startTime": "2020-09-30T16:06:03.703458Z"
			}, {
				"eventName": "queued",
				"durationInMicroSec": "74",
				"startTime": "2020-09-30T16:06:03.703466Z"
			}, {
				"eventName": "pipelined",
				"durationInMicroSec": "161",
				"startTime": "2020-09-30T16:06:03.703540Z"
			}, {
				"eventName": "transitTime",
				"durationInMicroSec": "587",
				"startTime": "2020-09-30T16:06:03.703701Z"
			}, {
				"eventName": "received",
				"durationInMicroSec": "74",
				"startTime": "2020-09-30T16:06:03.704288Z"
			}, {
				"eventName": "completed",
				"durationInMicroSec": "1",
				"startTime": "2020-09-30T16:06:03.704362Z"
			}],
			"rntbdRequestLengthInBytes": 413,
			"rntbdResponseLengthInBytes": 769,
			"requestPayloadLengthInBytes": 0,
			"responsePayloadLengthInBytes": 303,
			"channelTaskQueueSize": 1,
			"pendingRequestsCount": 0,
			"serviceEndpointStatistics": {
				"availableChannels": 1,
				"acquiredChannels": 0,
				"executorTaskQueueSize": 0,
				"inflightRequests": 1,
				"lastSuccessfulRequestTime": "2020-09-30T16:06:03.702Z",
				"lastRequestTime": "2020-09-30T16:06:03.701Z",
				"createdTime": "2020-09-30T16:06:01.271855Z",
				"isClosed": false
			}
		},
		"requestResponseTimeUTC": "2020-09-30T16:06:03.704396Z",
		"requestResourceType": "Document",
		"requestOperationType": "Read"
	}],
	"supplementalResponseStatisticsList": [],
	"addressResolutionStatistics": {},
	"regionsContacted": ["https://10.37.129.3:8081/"],
	"retryContext": {
		"retryCount": 0,
		"statusAndSubStatusCodes": null,
		"retryLatency": 0
	},
	"metadataDiagnosticsContext": {
		"metadataDiagnosticList": null
	},
	"serializationDiagnosticsContext": {
		"serializationDiagnosticsList": null
	},
	"gatewayStatistics": null,
	"systemInformation": {
		"usedMemory": "97392 KB",
		"availableMemory": "16679824 KB",
		"systemCpuLoad": "empty"
	},
	"clientCfgs": {
		"id": 0,
		"numberOfClients": 1,
		"connCfg": {
			"rntbd": "(cto:PT5S, rto:PT5S, icto:PT0S, ieto:PT1H, mcpe:130, mrpc:30)",
			"gw": "(cps:1000, rto:PT5S, icto:null, p:false)",
			"other": "(ed: true, cs: false)"
		},
		"consistencyCfg": "(consistency: Eventual, mm: true, prgns: [westus1,eastus1])"
	}
}

```

### CODE REVIEW

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/ConnectionPolicy.java [hunk 1]
```diff
@@ -498,4 +498,10 @@ public String toString() {
             ", maxRequestsPerConnection=" + maxRequestsPerConnection +
             '}';
     }
+
+    public String asDiagnostics() {
+        // TODO: cache this to avoid recomputation per diagnostics request
+        // TODO: only include essential configs
```
Comment (kirankumarkolli) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/ConnectionPolicy.java:504]: Will it be addressed now?

Comment (moderakh) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/ConnectionPolicy.java:504]: yes. once we close on the high level of the PR what information we want to capture, etc. I will address this within the scope of this PR.

**Review by kirankumarkolli — APPROVED** [review]

**Review by FabianMeiswinkel — APPROVED** [review]
LGTM

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/ConnectionPolicy.java (hunk 1 above)
Comment (moderakh) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/ConnectionPolicy.java]: done.

**Review by FabianMeiswinkel — APPROVED** [review]

**Comment by moderakh** [comment]
all tests against live prod passed on earlier commit. last commit only an unused method removed.

**Pull request merged**